File: example/high_level/follow_sin.cpp
#include <unitree/robot/go2/sport/sport_client.hpp>
#include <unitree/idl/go2/SportModeCmd_.hpp>
#include <unitree/common/thread/recurrent_thread.hpp>
#include <unitree/robot/go2/obstacles_avoid/obstacles_avoid_client.hpp>
#include <math.h>
#include <unistd.h>
#include <iostream>
#include <unitree/idl/ros2/PointCloud2_.hpp>

class Custom
{
public:
  Custom() {}
  void control();

  unitree::robot::go2::SportClient tc;

  int c = 0;
  float dt = 0.002; // 0.001~0.01
};

void printPosition(const std::array<float, 2>& position) {
    std::cout << "Position - X: " << position[0] << ", Y: " << position[1] << std::endl;
}

void Custom::control()
{
    c++;
    
    int32_t ret;

    float vx = 0.3;
    float delta = 0.06;
    static float count = 0;
    count += dt;
    std::vector<unitree::robot::go2::PathPoint> path;
    for (int i=0; i<30; i++) {
      unitree::robot::go2::PathPoint p;
      float var = (count + i * delta);
      p.timeFromStart = i * delta;
      p.x = vx * var;
      p.y = 0.6 * sin(M_PI * vx * var);
      p.yaw = 2*0.6 * vx * M_PI * cos(M_PI * vx * var);
      p.vx = vx;
      p.vy = M_PI * vx * (0.6 * cos(M_PI * vx * var));
      p.vyaw = - M_PI * vx*2*0.6 * vx * M_PI * sin(M_PI * vx * var);
      path.push_back(p);
    }
    std::cout << "Sending trajectory follow command, iteration: " << c << std::endl;

    ret = tc.TrajectoryFollow(path);
    if(ret == 0){
        std::cout << "TrajectoryFollow command was successful." << std::endl;
    } else {
        std::cout << "Call to TrajectoryFollow failed with error: " << ret << std::endl;
    }
}

int main(int argc, char** argv)
{
  unitree::robot::ChannelFactory::Instance()->Init(0);

  Custom custom;
  custom.tc.SetTimeout(10.0f);
  custom.tc.Init();

  unitree::common::ThreadPtr threadPtr = unitree::common::CreateRecurrentThread(custom.dt * 1000000, std::bind(&Custom::control, &custom));

  while (1)
  {
    sleep(10);
  }

  return 0;
}


/*int main(int argc, char **argv)
{
    if (argc < 2)
    {
        std::cout << "Usage: " << argv[0] << " networkInterface" << std::endl;
        return -1;
    }
    
    unitree::robot::ChannelFactory::Instance()->Init(0, argv[1]);

    // 스포츠 클라이언트 초기화
    unitree::robot::go2::SportClient sport_client;
    sport_client.SetTimeout(10.0f);
    sport_client.Init();

    const int totalDuration = 7; // 전체 지속 시간 15초

    for(int i = 1; i <= totalDuration; ++i){
        if(i <= 3) {
            // 처음 5초 동안 전진
            sport_client.Move(0.8, 0, 0);
        } else if(i <= 2) {
            // 다음 2초 동안 왼쪽으로 이동
            sport_client.Move(0, 0.8, 0);
        } else {
            // 나머지 시간 동안 다시 전진
            sport_client.Move(0.8, 0, 0);
        }
        sleep(1); // 1초 기다림
    }

    sport_client.StopMove(); // 이동 중지

    return 0;
}*/
/*
int main(int argc, char **argv)
{
    if (argc < 2)
    {
        std::cout << "Usage: " << argv[0] << " networkInterface" << std::endl;
        return -1;
    }
    
    unitree::robot::ChannelFactory::Instance()->Init(0, argv[1]);

    // Obstacles Avoid Client initialization and activation
    unitree::robot::go2::ObstaclesAvoidClient obstacles_avoid_client;
    obstacles_avoid_client.Init();
    obstacles_avoid_client.SwitchSet(true);

    // SportClient initialization
    unitree::robot::go2::SportClient sport_client;
    sport_client.SetTimeout(10.0f);
    sport_client.Init();

    bool obstacleDetected = false;
    int duration = 15; // total duration time

    for(int i = 0; i < duration; ++i){
        obstacles_avoid_client.SwitchGet(obstacleDetected);
        if(obstacleDetected){
            // if obstacle is detected, move to left
            sport_client.Move(0, 0.8, 0);
        }else{
            // else, move to forward
            sport_client.Move(0.8, 0, 0);
        }
        sleep(1);
    }

    sport_client.StopMove();

    return 0;
}*/

File: example_py/high_level/follow_sin.py

import sys
import time
import math
import asyncio
import threading

sys.path.append('/home/seungwan/Desktop/wrapping/lib/python/x86_64/')
import robot_interface as sdk

def main():
    # if len(sys.argv) < 2:
    #     print(f"Usage: {sys.argv[0]} networkInterface")
    #     return -1
    
    # Assuming that the ChannelFactory and Init methods are properly bound in Python
    # sdk.ChannelFactory.Instance().Init(0, sys.argv[1])
    chan = sdk.ChannelFactory.Instance()
    chan.Init(0, "enx7cc2c64bce6b")

    # Initialize the sport client, assuming the translated classes have the same functionality
    sport_client = sdk.SportClient(False)
    sport_client.SetTimeout(10.0)
    sport_client.Init()

    total_duration = 3  # Total duration is 3 seconds

    for i in range(1, total_duration + 1):
        if i <= 3:
            # Move forward for the first 3 seconds
            sport_client.Move(0.8, 0, 0)
        elif i > 3 and i <= 5:  # Corrected the logic to ensure this branch executes
            # Move left for the next 2 seconds
            sport_client.Move(0, 0.8, 0)
        else:
            # Move forward for the remaining time
            sport_client.Move(0.8, 0, 0)
        time.sleep(1)  # Wait for 1 second

    sport_client.StopMove()  # Stop moving

    return 0

if __name__ == "__main__":
    sys.exit(main())



'''
M_PI = 3.14159265358979323846

class Custom:
  c = 0
  dt = 0.002 ## 0.001~0.01

  def __init__(self):
    self.tc = sdk.SportClient(False)
    self.count = 0

  def control(self):
    self.c += 1
    
    vx = 0.3
    delta = 0.06
    self.count += self.dt

    path = []
    for i in range(30):
      p = sdk.PathPoint()
      var = (self.count + i * delta)
      p.timeFromStart = i * delta
      p.x = vx * var
      p.y = 0.6 * math.sin(M_PI * vx * var)
      p.yaw = 2*0.6 * vx * M_PI * math.cos(M_PI * vx * var)
      p.vx = vx
      p.vy = M_PI * vx * (0.6 * math.cos(M_PI * vx * var))
      p.vyaw = - M_PI * vx*2*0.6 * vx * M_PI * math.sin(M_PI * vx * var)
      path.append(p)


    ret = self.tc.TrajectoryFollow(path)
    print(f"{self.c}, {ret}") ## debugging
    if ret != 0:
      print(f"Call TrajectoryFollow: {ret}")

async def run_in_background(t ,func):
  while True:
    func()
    await asyncio.sleep(t)

def run_async_loop_in_thread(t, func):
    asyncio.run(run_in_background(t, func))

if __name__ == '__main__':

  chan = sdk.ChannelFactory.Instance()
  chan.Init(0, "")

  custom = Custom()
  custom.tc.SetTimeout(10.0)
  custom.tc.Init()
  t = threading.Thread(target=run_async_loop_in_thread, args=(1 * Custom.dt, custom.control))
  t.start()

  while True:
    time.sleep(10)
'''
